pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,290 @@
+		logger:   f.logger,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server. Here, os.FileMode is unused.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	if name == "" {
+		f.logger.Errorf("Mkdir Failed Provide a valid directory : %v", errEmptyDirectory)
+		return errEmptyDirectory
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.MakeDir(filePath)
+	if err != nil {
+		f.logger.Errorf("Mkdir Failed Error creating directory at %q : %v", filePath, err)
+		return err
+	}
+
+	f.logger.Logf("Mkdir Success %s successfully created", name)
+
+	return nil
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// The directories are not created if even one directory exist.
+func (f *ftpFileSystem) MkdirAll(path string, _ os.FileMode) error {
+	if path == "" {
+		f.logger.Errorf("MkdirAll Failed Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+
+	dirs := strings.Split(path, "/")
+
+	currentDir := dirs[0]
+
+	for i, dir := range dirs {
+		// Ignore empty directory names (can happen if there are double slashes).
+		if dir == "" {
+			continue
+		}
+
+		if i == 0 {
+			currentDir = dir
+		} else {
+			currentDir = fmt.Sprintf("%s/%s", currentDir, dir)
+		}
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			f.logger.Errorf("MkdirAll Failed Error : %v", err)
+			return err
+		}
+	}
+
+	f.logger.Logf("MkdirAll Success Directories creation completed successfully.")
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *ftpFileSystem) Open(name string) (File, error) {
+	if name == "" {
+		f.logger.Errorf("Open_file Failed Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Open_file Failed Error opening file : %v", err)
+		return nil, err
+	}
+
+	s := strings.Split(filePath, "/")
+
+	f.logger.Logf("Open_file Success Filepath : %q", filePath)
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+	}, nil
+}
+
+// permissions are not clear for Ftp as file commands do not accept an argument and don't store their file permissions.
+// currently, this function just calls the Open function. Here, os.FileMode is unused.
+func (f *ftpFileSystem) OpenFile(name string, _ int, _ os.FileMode) (File, error) {
+	return f.Open(name)
+}
+
+// Remove deletes a file from the FTP server.
+func (f *ftpFileSystem) Remove(name string) error {
+	if name == "" {
+		f.logger.Errorf("Remove_file Failed Provide a valid filename : %v", errEmptyFilename)
+		return errEmptyFilename
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.Delete(filePath)
+	if err != nil && fmt.Sprint(err) != transferComplete {
```

Comment:
Nope

Please assert the error is a textproto.Error by using errors.As

https://pkg.go.dev/net/textproto#Error

Then check the code is equal to StatusClosingDataConnection via

https://github.com/jlaffaye/ftp/blob/v0.2.0/status.go#L23

Reply:
thanks ....was stuck here.